amazon-kinesis-client/src/main/java/software/amazon/kinesis/coordinator/PeriodicShardSyncManager.java

```diff
@@ -275,22 +308,63 @@ private void runShardSync() {
      * @throws InvalidStateException
      */
     private Map<StreamIdentifier, List<Lease>> getStreamToLeasesMap(
-            final Set<StreamIdentifier> streamIdentifiersToFilter)
+            final Set<StreamIdentifier> streamIdentifiersToFilter, MetricsScope scope)
             throws DependencyException, ProvisionedThroughputException, InvalidStateException {
-        final List<Lease> leases = leaseRefresher.listLeases();
-        if (!isMultiStreamingMode) {
-            Validate.isTrue(streamIdentifiersToFilter.size() == 1);
-            return Collections.singletonMap(streamIdentifiersToFilter.iterator().next(), leases);
-        } else {
+        int numIncompatibleLeases = 0;
+        int numSingleStreamLeases = 0;
+        int numMultiStreamLeases = 0;
+        try {
+            final List<Lease> leases = leaseRefresher.listLeases();
             final Map<StreamIdentifier, List<Lease>> streamToLeasesMap = new HashMap<>();
-            for (Lease lease : leases) {
-                StreamIdentifier streamIdentifier = StreamIdentifier
+            if (isMultiStreamingMode) {
+                for (Lease lease : leases) {
+                    boolean isActiveLease = SentinelCheckpoint.SHARD_END.toString().equals(lease.checkpoint().sequenceNumber());
+                    if (lease instanceof MultiStreamLease) {
+                        numMultiStreamLeases += isActiveLease ? 1 : 0;
+                        StreamIdentifier streamIdentifier = StreamIdentifier
                         .multiStreamInstance(((MultiStreamLease) lease).streamIdentifier());
-                if (streamIdentifiersToFilter.contains(streamIdentifier)) {
-                    streamToLeasesMap.computeIfAbsent(streamIdentifier, s -> new ArrayList<>()).add(lease);
+                        if (streamIdentifiersToFilter.contains(streamIdentifier)) {
+                            streamToLeasesMap.computeIfAbsent(streamIdentifier, s -> new ArrayList<>()).add(lease);
+                        }
+                    } else {
+                        numSingleStreamLeases += isActiveLease ? 1 : 0;
+                        numIncompatibleLeases += isActiveLease ? 1 : 0;
```

Comment:
I dont understand the difference between the two. Also I think it would be better if we did the compatiblity check in scheduler. I believe this would be an uncommon case and doing it every iteration of PSSM is not efficient and also makes the code hard to understand and maintain.

Reply:
There are a couple of purposes: 
1. Ignore leases if stream identifier in configuration does not match stream identifier in the lease records. This is relevant for SingleStreamCompatible and SingleStreamUpgrade modes. For SingleStream and MultiStream modes, there's no change in behaviour. 
2. Generate periodic metrics for total number of leases in each format. 

I think (2) can be done in scheduler but would require a periodic scan of the table and a new metrics operation. It doesn't fit into the one operation scheduler is emitting metrics for, which is "MultistreamTracker". Happy to discuss that more. Otherwise, I am now thinking that maybe we should do (2) separately even if it stays in PSSM, to ensure that the original behavior in SingleStream and MultiStream modes remains the same.

Reply:
Have removed this logic after our discussion yesterday. Have moved the metrics to LeaseTaker.